cli/src/parse.rs

```diff
@@ -67,67 +63,84 @@ fn output_file_name(language_type: SupportedLanguage, crate_name: &CrateName) ->
                 import_map
             },
         )
 }
 
-/// Collect all the parsed sources into a mapping of crate name to parsed data.
-pub fn parse_input(
-    inputs: impl ParallelIterator<Item = ParserInput>,
+fn parse_dir_entry(
+    parse_context: &ParseContext,
+    language_type: SupportedLanguage,
+    dir_entry: &DirEntry,
+) -> Result<Option<ParsedData>, ParseError> {
+    if dir_entry.path().is_dir() {
+        return Ok(None);
+    }
+
+    let Some(parse_file_context) =
+        parse_file_context(parse_context.multi_file, language_type, dir_entry)
+            .map_err(|err| ParseError::IOError(err.to_string()))?
+    else {
+        return Ok(None);
+    };
+
+    typeshare_core::parser::parse(parse_context, parse_file_context)
+}
+
+/// Use parallel builder to walk all source directories concurrently.
+pub fn parallel_parse(
     parse_context: &ParseContext,
+    walker_builder: WalkBuilder,
+    language_type: SupportedLanguage,
 ) -> anyhow::Result<BTreeMap<CrateName, ParsedData>> {
-    inputs
-        .into_par_iter()
-        .try_fold(
-            BTreeMap::new,
-            |mut parsed_crates: BTreeMap<CrateName, ParsedData>,
-             ParserInput {
-                 file_path,
-                 file_name,
-                 crate_name,
-             }| {
-                let parse_file_context = ParseFileContext {
-                    source_code: std::fs::read_to_string(&file_path)
-                        .with_context(|| format!("Failed to read input: {file_name}"))?,
-                    crate_name: crate_name.clone(),
-                    file_name: file_name.clone(),
-                    file_path,
-                };
-
-                let parsed_result =
-                    typeshare_core::parser::parse(parse_context, parse_file_context)
-                        .with_context(|| format!("Failed to parse: {file_name}"))?;
-
-                if let Some(parsed_data) = parsed_result {
-                    parsed_crates
-                        .entry(crate_name)
-                        .or_default()
-                        .add(parsed_data);
-                }
+    let (tx, rx) = bounded::<anyhow::Result<ParsedData>>(100);
 
-                Ok(parsed_crates)
-            },
-        )
-        .try_reduce(BTreeMap::new, |mut file_maps, parsed_crates| {
-            for (crate_name, parsed_data) in parsed_crates {
-                file_maps.entry(crate_name).or_default().add(parsed_data);
+    let collector_thread = thread::spawn(move || {
+        let mut crate_parsed_data: BTreeMap<CrateName, ParsedData> = BTreeMap::new();
+
+        for result in rx {
+            let parsed_data = result?;
+            let crate_name = parsed_data.crate_name.clone();
+            // Append each yielded parsed data by its respective crate.
+            *crate_parsed_data.entry(crate_name).or_default() += parsed_data;
+        }
+
+        Ok(crate_parsed_data)
+    });
+
+    walker_builder.build_parallel().run(|| {
```

Comment:
does this outer closure run exactly 1 time? If no, then I think it would be better to create the boxed closure outside. If yes then disregard

Reply:
The outer closure is called on each dir entry yielded. The inner for each thread.

https://docs.rs/ignore/latest/ignore/struct.WalkParallel.html#method.run

Similar to parallel walker example: https://github.com/BurntSushi/ripgrep/blob/master/crates/ignore/examples/walk.rs#L28.